compiler/rustc_mir/src/transform/check_unsafety.rs

```diff
@@ -232,37 +232,40 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
                     UnsafetyViolationKind::GeneralAndConstFn,
                     UnsafetyViolationDetails::DerefOfRawPointer,
                 ),
-                ty::Adt(adt, _) => {
-                    if adt.is_union() {
-                        if context == PlaceContext::MutatingUse(MutatingUseContext::Store)
-                            || context == PlaceContext::MutatingUse(MutatingUseContext::Drop)
-                            || context == PlaceContext::MutatingUse(MutatingUseContext::AsmOutput)
-                        {
-                            let elem_ty = match elem {
-                                ProjectionElem::Field(_, ty) => ty,
-                                _ => span_bug!(
-                                    self.source_info.span,
-                                    "non-field projection {:?} from union?",
-                                    place
-                                ),
-                            };
-                            if !elem_ty.is_copy_modulo_regions(
+                ty::Adt(adt, _) if adt.is_union() => {
+                    if context == PlaceContext::MutatingUse(MutatingUseContext::Store)
+                        || context == PlaceContext::MutatingUse(MutatingUseContext::Drop)
+                        || context == PlaceContext::MutatingUse(MutatingUseContext::AsmOutput)
```

Comment:
What slightly concerns me here is that this is the `context` of the entire place, but we are considering here just a single place projection. What if there is a deref involved?

```rust
#![feature(untagged_unions)]

union U {
    p: &'static mut i32,
}

fn test(u: U) {
    *(u.p) = 13;
}
```
This does error, but the error claims "assignment to non-`Copy` union field", which is not really what happens (the value of `p` is not changed). The problem is that this is reading a `union` field. If I made the test here even smarter to take into account `needs_drop`, then that code above would be incorrectly accepted!

In fact, union fields that drop are not even permitted any more (with no amount of feature gates).

Reply:
Fixed this by checking for `Deref` explicitly. The logic is kind of strange now but I think it works -- and I added a test to confirm that the test above how fails with the right error:
```
error[E0133]: access to union field is unsafe and requires unsafe function or block
  --> $DIR/union-unsafe.rs:31:5
   |
LL |     *(u.p) = 13;
   |     ^^^^^^^^^^^ access to union field
   |
   = note: the field may not be properly initialized: using uninitialized data will cause undefined behavior
```
(instead of "assignment to union field")